new: :sparkles: add ASGI middleware alternative

thentgesMindee opened the pull request:
fix https://github.com/laurentS/slowapi/issues/98

Might be easier for people reviewing to explain the different commits:
- first commit adds the new ASGI middleware
- second one update tests so that it test both the HTTP and ASGI middlewares (using pytest parametrize and fixtures)
- third one is a refactoring commit, to avoid duplicating too much code between the HTTP and the ASGI middleware

Might be easier to first review the first two commits, then the refactoring one (to avoid seeing too many changes at once)

Let me know what do you think about this :smile:

Note: maybe I could do a bit of refactoring  between `_inject_headers` and `_inject_asgi_headers` as well ?

laurentS commented:
Thanks for this PR @thentgesMindee ! Just to say I have no time to look into it this week (nor next, most likely), but there are other contributors to this repo who might get to it before then.

thentgesMindee commented:
@laurentS alright thanks a lot for the information, no worries :smile:
I see you enabled the github actions check on the branch and they fail because some changes were made on master and not rebased on my branch, gonna fix this right away

thentgesMindee commented:
rebase made, + commit to update newly added tests
all checks should pass this time :smile:

laurentS commented on slowapi/middleware.py:
```diff
@@ -54,3 +58,89 @@ async def dispatch(
+        if handler is None:
+            return await self.app(scope, receive, self.send)
+
+        name = "%s.%s" % (handler.__module__, handler.__name__)
```
Any particular reason not to use the f-string notation here? I thought it was faster pretty much all the time. Personally, I also find it much more readable.

thentgesMindee replied:
not really, just a refactor of the existing code, didn't touch this line

laurentS commented on slowapi/middleware.py:
```diff
@@ -1,56 +1,152 @@
+                type(e), _rate_limit_exceeded_handler
+            )
+            return exception_handler(request, e), False
+        # request.state._rate_limiting_complete = True
```
Can we remove this?

laurentS commented on slowapi/middleware.py:
```diff
@@ -1,56 +1,152 @@
+    return None, False
+
+
+def _should_exempt(limiter: Limiter, handler) -> bool:
```
```suggestion
def _should_exempt(limiter: Limiter, handler: Optional[Callable]) -> bool:
```

laurentS commented on slowapi/middleware.py:
```diff
@@ -1,56 +1,152 @@
+
+
+def _check_limits(
+    limiter: Limiter, request: Request, handler, app: Starlette
```
```suggestion
    limiter: Limiter, request: Request, handler: Optional[Callable], app: Starlette
```

laurentS commented on slowapi/middleware.py:
```diff
@@ -1,56 +1,152 @@
+    return handler
+
+
+def _get_route_name(handler):
```
```suggestion
def _get_route_name(handler: Callable) -> str:
```

thentgesMindee replied:
this would raise some linting errors `cannot find __module__ in (...) -> Any`, so I used Type[Callable] instead.

laurentS commented on slowapi/middleware.py:
```diff
@@ -1,56 +1,152 @@
 from slowapi import Limiter, _rate_limit_exceeded_handler
 
 
+def _find_route_handler(routes: Iterable[BaseRoute], scope: Scope):
```
```suggestion
def _find_route_handler(routes: Iterable[BaseRoute], scope: Scope) -> Optional[Callable]:
```

laurentS commented on slowapi/middleware.py:
```diff
@@ -1,56 +1,152 @@
+from typing import Optional, Iterable, Tuple
```
```suggestion
from typing import Callable, Iterable, Optional, Tuple
```

laurentS reviewed (changes requested):
Sorry for the delay in reviewing this. I suggested a few type annotations, hoping to at least get `mypy` to look into each function.
I think it'd be nice to give a minimal example of each usage in the docs as well, and maybe add a line explaining how you'd choose one or the other middleware.

thentgesMindee commented:
Hi thanks for the review @laurentS !
I spotted a bug in my middleware, I'll fix it then handle your review requests

thentgesMindee commented on slowapi/middleware.py:
```diff
@@ -54,3 +58,89 @@ async def dispatch(
+        # Limit
+        request = Request(scope, receive=receive, send=self.send)
+
+        # let the decorator handle if already in
```
pretty much a copy/paste from what was in the HTTP middleware, let's remove it `_should_exempt` function

thentgesMindee commented:
Just as an update, committed the fix I was talking about, as well as type annotations change requests.
will do some more commits either today or tomorrow for docstring and documentation :ok_hand:

thentgesMindee commented:
Sorry for the delay, but I finally managed to find some time to:
- add support for both asynchronous or synchronous error handlers in the ASGI Middleware
- add a small documentation about choosing between ASGI and WSGI middleware
- some docstrings

Let me know what do you think about it

thentgesMindee commented:
Hi @laurentS , sorry for being impatient, any chance you could review the requested changes ?
Don't want to rush you, just want to make sure this is not lost in time

laurentS commented:
@thentgesMindee sorry for the lag. Things are a bit busy everywhere. I just approved the CI run, hadn't noticed this was waiting for a button click, and it failed on typing. Can you check and correct?

thentgesMindee commented:
Sure no worries take your time, don't want to rush you, as I said I just wanted to make sure the PR wasn't forgotten.
Will have a look to the failure and ping you back when I resolve it

thentgesMindee commented:
@laurentS type annotations were updated (just left a # type: ignore, but the other ones are fixed)

thentgesMindee commented:
Just saw that the CI is still failing because of some unused imports, gonna fix it tomorrow!

laurentS reviewed (dismissed):
Looks good to me. Just a suggestion in `_get_route_name` to use a more modern style, otherwise all LGTM!

thentgesMindee commented:
Thanks a lot for the review! rebased my branch to be up-to-date with current `master`, and added the last change about `_get_route_name`. All tests are passing.
Could you re-approve (can't approve my own PR, and seems that my force-push reset the old approvals) so that we merge this ?
Thanks again :pray:
